Adds migrations and sets up AccountLink APIs

jayeshsidhwani opened the pull request:
- adds support for flyway and jooq
- adds migrations for Account and Consent
- adds APIs for Account Linking

Commit:
Sets up base models and database migrations
- adds support for flyway and jooq
- adds migrations for Account and Consent
- adds a pre-save trigger to update updated_at column

Commit:
- wires AccountDiscover request / response

Commit:
Wires Discover response
- adds AccountDao
- sets up DiscoverResponse using DiscoverRequest and Account

Commit:
Sets up multiple Account objects based on fiTypes

Commit:
- Changes the API structure to comply with ReBit recommendations here: https://swagger-ui.rebit.org.in/?url=https://specifications.rebit.org.in/api_specifications/account_aggregator/FIP_1_1_1.yaml
- Set up AccountLink API

JaiNanavati commented on src/main/java/org/indiastack/depa/provider/models/account/Account.java:
```diff
@@ -54,7 +54,7 @@ public Account(String linkRefNumber,
     public Account(DiscoverRequest discoverRequest, String fiType) {
         this.status = AccountStatus.INITIATED;
```
possible values for status enum are 'DISCOVERED', 'DELINKED','LINKED'

JaiNanavati commented on src/main/java/org/indiastack/depa/provider/controllers/v1/AccountController.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @GetMapping("/link/{RefNumber}/{Token}")
+    public void getLinkedAccount() {
```
do we need to specific PathVariables for Refnumber and token in the constructor args?

JaiNanavati commented on src/main/java/org/indiastack/depa/provider/controllers/v1/ConsentController.java:
```diff
@@ -1,12 +1,12 @@
 
 @RestController
-@RequestMapping("v1/gateway/consents")
+@RequestMapping("Consent")
```
Do we need to stick with caps here? I know api specs show caps right now but lets talk about this

jayeshsidhwani replied:
we spoke - following the Rebit contract now

JaiNanavati commented on src/main/java/org/indiastack/depa/provider/models/account/Account.java:
```diff
@@ -0,0 +1,74 @@
+    private String customerAAId;
+
+    // @todo: which accountNumber is this?
+    private String accountNumber;
```
This is the users account number of the account they want to link

JaiNanavati commented on src/main/java/org/indiastack/depa/provider/models/account/Account.java:
```diff
@@ -0,0 +1,74 @@
+    private String accountNumber;
+
+    // @todo: which accountRefNumber is this?
+    private String accountRefNumber;
```
This is bank metadata around the account number. Need to clarify this and will take on action item.

JaiNanavati commented on src/main/java/org/indiastack/depa/provider/models/account/Account.java:
```diff
@@ -0,0 +1,74 @@
+        this.status = AccountStatus.INITIATED;
+    }
+
+    public Account(DiscoverRequest discoverRequest, String fiType) {
```
Is there a reason fiType is outside DiscoverRequest ? Also, current request stipulates an array. My understanding is the following: a particular account can have multiple fiTypes within it (Deposit, Withdrawal etc.). Shouldnt stipulations on data to retrieve from an account be Consent specific as opposed to something tied to account linkage?

jayeshsidhwani replied:
Changed this based on our discussion on call

JaiNanavati commented on src/main/java/org/indiastack/depa/provider/models/account/Account.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    public Account(DiscoverRequest discoverRequest, String fiType) {
+        this.status = AccountStatus.INITIATED;
```
Think ive commented elsewhere, but INITIATED not a valid state currently

JaiNanavati commented on src/main/java/org/indiastack/depa/provider/models/account/http/DiscoverResponse.java:
```diff
@@ -0,0 +1,29 @@
+
+@AllArgsConstructor
+@Getter
+public class DiscoverResponse extends BaseRequest {
```
Should we be creating a new BaseResponse class for response related classes?

jayeshsidhwani replied:
done

Commit:
Dummy implementation for discover account
- adds support for JooQ as ORM
- implements AccountDao for bulkInsert operation
- all Response objects now implement BaseResponse
- refactors Account fields based on PR discussion
- adds correct enum for AccountType, AccountStatus, FiType, and ConsentType
- moved implementation of all enums to source code instead of database

Commit:
- adds dummmy implementation for account link deletion and fetch

JaiNanavati commented on src/main/java/org/indiastack/depa/provider/services/AccountService.java:
```diff
@@ -0,0 +1,48 @@
+    public LinkResponse linkAccount(LinkRequest linkRequest) {
+        // Get all accounts from the database
+        // Update link status to LINKED
+        // Generate Token?? @Todo
```
We will need a `confirmLink(String refNumber, String token)` function in which to update the link status to LINKED.

JaiNanavati commented on src/main/java/org/indiastack/depa/provider/services/AccountService.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public LinkResponse linkAccount(LinkRequest linkRequest) {
+        // Get all accounts from the database
```
With respect to the token, since we are validating we will definitely have to call a token generator and probably store it alongside individual accounts in the table as well.
